src/openassetio-python/PyRetainingSharedPtr.hpp

```diff
@@ -0,0 +1,250 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 The Foundry Visionmongers Ltd
+#pragma once
+/**
+ * Defines PyRetainingSharedPtr, a custom `shared_ptr` that keeps the Python
+ * instance alive whilst the associated C++ instance is alive.
+ *
+ * This works around https://github.com/pybind/pybind11/issues/1333
+ *
+ * The solution was taken and modified from
+ * https://github.com/pybind/pybind11/issues/1546
+ */
+#include <memory>
+#include <utility>
+
+#include <pybind11/pybind11.h>
+
+#include <openassetio/export.h>
+
+namespace openassetio {
+inline namespace OPENASSETIO_CORE_ABI_VERSION {
+/**
+ * Custom shared_ptr type for flagging function arguments and trampoline
+ * return values that should keep the Python object alive at least as
+ * long as the shared_ptr (and any other shared_ptrs copy-constructed
+ * from it) is alive.
+ *
+ * This smart pointer type should be used as the parameter type for
+ * functions that take a Python object that inherits from and overrides
+ * virtual methods of a C++ class (so only those classes that have a
+ * pybind "trampoline").
+ *
+ * Similarly, this should be used as the return type in any pybind
+ * "trampoline" override (i.e. PYBIND_OVERRIDE et al) that returns a
+ * base class pointer where the derived class is implemented in Python.
+ *
+ * Without this we risk "RuntimeError: Tried to call pure virtual
+ * function" if the Python object goes out of scope but a C++ object
+ * still holds a shared_ptr to (the C++ base class of) it.
+ *
+ * @tparam T Wrapped type.
+ */
+template <class T>
+struct PyRetainingSharedPtr : std::shared_ptr<T> {
+  using std::shared_ptr<T>::shared_ptr;
+};
+
+/**
```

Comment:
[fix] Would it be possible to pop the helpers in a separate Doxygen group (even though we dont scan python atm), and add some examples + documentation on how all this works. It is very neat indeed, but particularly opaque.